src/cargo/util/flock.rs

```diff
@@ -193,57 +195,78 @@ impl Filesystem {
     where
         P: AsRef<Path>,
     {
-        self.open(
-            path.as_ref(),
-            OpenOptions::new().read(true),
-            State::Shared,
-            config,
-            msg,
-        )
+        let (path, f) = self.open(path.as_ref(), &OpenOptions::new().read(true), false)?;
+        self.lock(path, f, config, msg, false)
     }
 
-    fn open(
+    /// Opens shared access to a file, returning the locked version of a file.
+    ///
+    /// Compared to [`Filesystem::open_ro`], this will create the file (and
+    /// any directories in the parent) if the file does not already exist.
```

Comment:
Why this difference?

Reply:
The first time locking the package cache, it needs to create the lock file in order to lock it.

The `open_ro` variant is used with locking `Cargo.lock`, where it doesn't want to create an empty file if `Cargo.lock` doesn't exist.

So, one is used for locking when there isn't anything relevant to lock against (like the package cache), and the other is for locking pre-existing files (like `Cargo.lock`).

Reply:
If that is the main distinction, I feel like the name should convey that rather than trying to push the "shared" policy naming down to this level.

Reply:
Sorry, I'm not quite sure what else would convey that better, or I'm not understanding your concern. It is opening a file with a shared lock, creating it if it doesn't exist. The name seems to be describing exactly what it does, so I'm not clear what it could be instead?

Or are you referring to `open_ro`? I could imagine that being `open_shared_ro` to make it more explicit what it is doing (though read-only implies it is shared).

Reply:
The lack of consistency in the naming was drawing my attention to the wrong things, making this harder to understand.

One idea
- `open_shared`
- `create_shared` or `create_or_open_shared` (some `create`s are idempotent while others aren't)
- `create_exclusive` or `create_or_open_exclusive`

Reply:
I decided to go with the following:

| Before | After |
|--------|-------|
| open_rw | open_rw_exclusive_create |
| try_open_rw | try_open_rw_exclusive_create |
| open_ro | open_ro_shared |
| open_shared_create | open_ro_shared_create |
| try_open_shared_create | try_open_ro_shared_create |

The form is roughly (`try_`) `open_` *access_mode* *lock_mode* (`_create`).

The intent is to be explicit about exactly what the method is doing. I'm not a huge fan of wordy methods, but I think each word is important to understand what the method is doing. An alternate solution might be to use a builder (similar to `OpenOptions`), but with only 5 variants, I'm not sure that is necessary at this stage. If there were more, I would advocate for that.

Reply:
Not the biggest fan but it works